[4.0] Debug bar

elkuku opened the pull request:
This is a redo of #15472 

### Summary of Changes

The goal of this PR is to refactor the debug plugin (currently more than 2.000 lines of code) into smaller pieces.
An external library is used to display a floating debug bar. [See a demo](http://phpdebugbar.com/)

### Testing Instructions
Run

    composer install
    npm i

Turn on debug and/or language debug and play with the various options in plugin configuration.

### Expected result

#### A beautiful floating debug toolbar (mobile ready) :wink: 
![bildschirmfoto_2018-05-12_16-19-09](https://user-images.githubusercontent.com/33978/39961654-bc5d9e66-5600-11e8-80d3-7e7f79ab08d0.png)

### Actual result

#### A not so beautiful looong thingy, fixed to the very bottom of the page.

![bildschirmfoto_2018-05-12_16-20-57](https://user-images.githubusercontent.com/33978/39961655-bca736b6-5600-11e8-821a-6ade307f0973.png)

### Documentation Changes Required
Alot :tongue: 

### Extended Description

While working on a [Symfony](https://symfony.com) project, I was amazed by the debug bar they offer and I thought about the old Joomla! plugin...<br />
Integrating the Symfony bar would require a lot of dependencies from Symfony, probably including the Symfony Application class. While this would be amazing, I suspect that this is not gonna happen any time soon...

So I started playing with [maximebf/debugbar](https://github.com/maximebf/php-debugbar) which claims to be inspired by the Symfony debug bar, but for "non Symfony" projects `;)`

This resulted in some `DataCollector` classes which, apart from collecting data, are responsible for bringing the collected data to something "serializable".

This data is then serialized and written to disk for later analysis. (**Not implemented yet**)

After a thousand words, here are some screens:
**Note**: Since the "old" code is still in place, you will see the "old" output as well as the "new " output in the screenshots.

----

### Session
![2017-04-18-121246_1366x768_scrot](https://cloud.githubusercontent.com/assets/33978/25298253/2494a056-26b9-11e7-9665-c3a25c06b794.png)

----

### Profile
![2017-04-21-133031_1366x768_scrot](https://cloud.githubusercontent.com/assets/33978/25298255/249a030c-26b9-11e7-8b95-0a6abbf31da6.png)

----

### Queries
The database query tab features including
    * Highlighted queries
    * Timings and memory reports
    * A (clickable) call stack with indication of the "caller" (somewhat experimental)
    * Duplicated query check
    * Duplicated queries can be filtered **Note** the filter buttons are at the lower right corner :wink: 
    * A button to "copy" the query to the clipboard

![bildschirmfoto_2018-09-02_12-34-16](https://user-images.githubusercontent.com/33978/44958911-9dcf0c00-aeac-11e8-9dd9-312d897784cd.png)

Call stack
![joomla 4 debug bar -call-stack](https://user-images.githubusercontent.com/33978/44042430-dc38f482-9ee5-11e8-91ec-9bd2667c322a.png)

Duplicates filter
![menus items main menu joomla 4 debug bar administration](https://user-images.githubusercontent.com/33978/44042438-dfd1afd0-9ee5-11e8-9770-2dcbccc82443.png)

Explains
![bildschirmfoto_2018-09-04_11-39-58](https://user-images.githubusercontent.com/33978/45045025-722c5d00-b037-11e8-9f98-2d914a30f273.png)

----

### Deprecation logging
![2017-04-21-171552_1366x768_scrot](https://cloud.githubusercontent.com/assets/33978/25298260/24d1f938-26b9-11e7-90b1-7e585fa01473.png)
![bildschirmfoto_2018-05-12_16-37-52](https://user-images.githubusercontent.com/33978/39961780-ff001c24-5602-11e8-8a8c-9503f0d5d91a.png)

----

### Language stuff
Should look as before, just more readable and compact.

#### Loaded

![bildschirmfoto_2018-09-02_11-08-43](https://user-images.githubusercontent.com/33978/44958113-fac4c500-aea0-11e8-870b-46f0d0c2812e.png)

#### Untranslated

![bildschirmfoto_2018-09-02_11-09-44](https://user-images.githubusercontent.com/33978/44958114-fb5d5b80-aea0-11e8-8992-ef1644259c0c.png)

#### Errors 

![bildschirmfoto_2018-09-02_11-10-42](https://user-images.githubusercontent.com/33978/44958115-fb5d5b80-aea0-11e8-8517-11abd55f82c2.png)

----

### Mobile ready
![screenshot_20170426-010917](https://cloud.githubusercontent.com/assets/33978/25421343/054e2354-2a22-11e7-8501-256bd4e49acc.png)

Tested on a 5'' screen with fat fingers and it worked pretty well.

----

Happy debugging `=;)`

# P.S.: `@todo` Things that should be added (later)

* [x] Memory display on profile tab - Thanks @Hackwar 
* [ ] A lot of visual improvements to be implemented using modern CSS technioques - Thanks `@all`

elkuku committed:
Add vendor files

elkuku committed:
Add data collectors

elkuku committed:
Add media files

elkuku committed:
Add vendor media files

elkuku committed:
Update debug plugin

elkuku committed:
Update language file

elkuku committed:
Update debug plugin

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

elkuku committed:
Remove vendor media files

elkuku committed:
Update .gitignore

brianteeman commented on administrator/language/en-GB/en-GB.plg_system_debug.ini:
```diff
@@ -21,6 +21,8 @@ PLG_DEBUG_FIELD_LOG_CATEGORIES_LABEL="Log Categories"
 PLG_DEBUG_FIELD_LOG_CATEGORY_MODE_EXCLUDE="Exclude"
 PLG_DEBUG_FIELD_LOG_CATEGORY_MODE_INCLUDE="Include"
 PLG_DEBUG_FIELD_LOG_CATEGORY_MODE_LABEL="Log Category Mode"
+PLG_DEBUG_FIELD_LOG_DEPRECATED_CORE_DESC="Only relevant for Joomla! core developers"
```
we dont need this string, just remove it please.

elkuku replied:
Done :wink:

elkuku committed:
Remove a language string - thanks @brianteeman

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

elkuku committed:
Update debug.xml

laoneo commented:
Does the query tab has duplicate query detection and can show the stack trace? That's two very useful features of the current debug plugin.

mbabker commented:
Yes it can highlight duplicate queries.  I don't remember off hand if the debug bar natively supports collecting stack traces, but the Laravel package that extends this does have them so even if it's not a native capability it can be added in.

mbabker commented:
Actually, we couldn't use the existing database integrations in the main debug bar package anyway.  That one only supports PDO, we have non-PDO drivers in our database API (MySQLi and SQL Server).

elkuku commented:
> Does the query tab has duplicate query detection and can show the stack trace? That's two very useful features of the current debug plugin.

Currently only the executed queries are displayed in a custom tab. For the reasons Michael stated we can not use the provided db tab so... - no stack trace, timings, memory etc.
I'd like to propose adding a component that displays additional information like this - in the next PR :wink:

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

# Conflicts:
#	composer.lock

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

# Conflicts:
#	plugins/system/debug/debug.php

elkuku committed:
Use __DEPLOY_VERSION__
Still learning...

ggppdk commented:
@elkuku 

hi, will this PR have solution for issue:
#21131

brianteeman commented:
@ggppdk why don't you test it and see

Hackwar commented:
Applying this to the current 4.0-dev branch fails on several levels. @elkuku can you fix the conflicts here?

elkuku commented:
> hi, will this PR have solution for issue:

e.g. will it speed up stuff?
At the current stage: no - because the "old" code processing db info is still in place - for reference.

My proposal would be to display only the executed queries in the debug bar and move the "extended information" (what we have currently) to a proper page that is displayed by a component - along with all the other information that we store about a request.
On the other hand - do we really need all the information or should we better rely on external tools for database profiling?

> can you fix the conflicts here?

Will do..

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

# Conflicts:
#	composer.json
#	libraries/vendor/composer/autoload_classmap.php
#	libraries/vendor/composer/autoload_files.php
#	libraries/vendor/composer/autoload_psr4.php
#	libraries/vendor/composer/autoload_static.php

elkuku committed:
Composer public files

elkuku committed:
Composer vendor and autoload files

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

elkuku committed:
Sooooo Let's register our namespace on our own... again :(

elkuku commented:
So... I managed to get this back to a workable state...

Apparently something has changes the way autoloading magic happens now so - I registered my namespace again by hand... - Maybe @wilsonge could help me out here again?

elkuku committed:
Merge branch '4.0-dev' into debug-bar-2

roland-d commented:
@elkuku Thank you for your work on this, it is looking very promising.

Here are my initial findings:

1. When I click the info link, it opens a new tab with a 404 that the profiler component is not found.
<img width="270" alt="image" src="https://user-images.githubusercontent.com/359377/43308346-9d1665b2-9181-11e8-9ec8-c11c1a16f7c5.png">

I cannot find where I should install this component as Discover doesn't see it.

2. I do not know if this is by design but the bars run through the text
<img width="321" alt="image" src="https://user-images.githubusercontent.com/359377/43308447-ec97e9c6-9181-11e8-9f4c-1b7029ac16c5.png">

3. The queries start counting at 0 instead of 1
<img width="487" alt="image" src="https://user-images.githubusercontent.com/359377/43308481-05d0ed02-9182-11e8-8bf4-f5afc65b6cfe.png">

4. In the queries we can no longer see if they took a long time or not, this makes it hard to find out which query is slow. We have this bar in the current debugger:
<img width="1368" alt="image" src="https://user-images.githubusercontent.com/359377/43308829-1b332236-9183-11e8-95d6-efe45d6f1e99.png">

You can click on a position to jump to the query. This is something we should keep.

6. Listing duplicate queries is useful too:
<img width="186" alt="image" src="https://user-images.githubusercontent.com/359377/43308859-3bd89502-9183-11e8-9e10-deaa01288efc.png">

7. In the list of loaded language files it would be great if the file links are clickable like in an XDebug trace. This allows the file to be opened in the IDE.

<img width="648" alt="image" src="https://user-images.githubusercontent.com/359377/43308882-55dc6fd2-9183-11e8-9f34-9ecd1053a5a5.png">

The link here goes to `phpstorm://open?file=/usr/local/var/www/rsform/plugins/system/pwtseo/language/en-GB/en-GB.plg_system_pwtseo.ini&line=`

8. In the Untranslated part, it only shows partial filenames
<img width="214" alt="image" src="https://user-images.githubusercontent.com/359377/43308922-7766ce54-9183-11e8-91c3-a912026b22c1.png">


9. In your screenshot you have a Deprecated tab, I don't see that. What is needed to trigger that?

10. Keep up the good work :)

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

# Conflicts:
#	.gitignore
#	libraries/vendor/composer/autoload_classmap.php
#	libraries/vendor/composer/autoload_files.php
#	libraries/vendor/composer/autoload_psr4.php
#	libraries/vendor/composer/autoload_static.php
#	libraries/vendor/composer/installed.json

elkuku committed:
Merge branch 'debug-bar-2' of https://github.com/elkuku/joomla-cms into debug-bar-2

elkuku committed:
Remove files from libraries/vendor

elkuku committed:
Move Files from media/vendor so they don't get removed on install...
@TODO remove and move properly

elkuku committed:
Change the path where our 3pd files live for now

mbabker commented:
> I do not know if this is by design but the bars run through the text

Yes, that's the design.  I use this same debug bar on the Joomla Framework website for local dev, see https://user-images.githubusercontent.com/368545/36637903-b2229e60-19ab-11e8-9b64-227b812f37f1.png for an example from there.  Basically, the time it starts measuring timings for the timeline is the left edge of the window and the time it stops is the right edge, so the bar positioning is somewhat relative to when in the request the timing was captured (and as you can see in that screenshot the timings aren't restricted to a single path, you can start and stop timing measurements a bit more granularly and catch different steps in a "higher" level topic such as the sub-tasks performed by a controller).

elkuku committed:
Copy debug bar assets using the build.js script

elkuku committed:
Remove vendor files

elkuku committed:
Merge branch '4.0-dev' into debug-bar-2

elkuku committed:
Ignore minified files

elkuku committed:
Revert Ignore minified files

elkuku committed:
Add minified files

elkuku committed:
Add some color
So somebody can scream at me =;)

elkuku committed:
Let's fetch the files from the 'right' place...

elkuku committed:
Update some code

elkuku committed:
Let's use tabs.....

elkuku committed:
Let's use tabs.....

Quy commented on plugins/system/debug/DataCollector/ProfileCollector.php:
```diff
@@ -0,0 +1,309 @@
+	public function stopMeasure($name, array $params = array())
+	{
+		$end = microtime(true);
+		if (!$this->hasStartedMeasure($name))
```
Add blank line above.

Quy commented on plugins/system/debug/DataCollector/ProfileCollector.php:
```diff
@@ -0,0 +1,309 @@
+		{
+			throw new DebugBarException("Failed stopping measure '$name' because it hasn't been started");
+		}
+		$this->addMeasure(
```
Add blank line above.

Quy commented on plugins/system/debug/debug.xml:
```diff
@@ -217,6 +168,55 @@
+						class="switcher"
+						label="PLG_DEBUG_FIELD_LOGS_LABEL"
+						default="1"
+				>
```
Indent.

Quy commented on plugins/system/debug/debug.xml:
```diff
@@ -217,6 +168,55 @@
 				label="PLG_DEBUG_LOGGING_FIELDSET_LABEL"
 				>
+				<field
+						name="logs"
```
Remove 1 indent from lines 172-176

Quy commented on plugins/system/debug/debug.xml:
```diff
@@ -217,6 +168,55 @@
+				</field>
+
+				<field
+						name="log_priorities"
```
Remove 1 indent for lines 183-187. Indent on line 188.

Quy commented on plugins/system/debug/debug.xml:
```diff
@@ -217,6 +168,55 @@
+				</field>
+
+				<field
+						name="log_categories"
```
Remove 1 indent for lines 201-206

Quy commented on plugins/system/debug/debug.xml:
```diff
@@ -217,6 +168,55 @@
+				/>
+
+				<field
+						name="log_category_mode"
```
Remove 1 indent for lines 210-214. Indent line 215.

Quy commented on plugins/system/debug/debug.xml:
```diff
@@ -228,6 +228,17 @@
+					class="switcher"
+					label="PLG_DEBUG_FIELD_LOG_DEPRECATED_CORE_LABEL"
+					default="0"
+				>
```
Indent lines 237-239.

elkuku committed:
Fix CS - Thank @Quy

elkuku committed:
Fix more CS

Quy commented on plugins/system/debug/DataCollector/ProfileCollector.php:
```diff
@@ -0,0 +1,312 @@
+	 * @param   array        $params     Parameters.
+	 * @param   string|null  $collector  A collector.
+	 *
+	 * @since __DEPLOY_VERSION__
```
Please align the second column

elkuku replied:
Sorry, not sure what you mean.. You are referring to the `@since` tag?

elkuku replied:
@Quy maybe you know what the current best practice is to run phpcs on the repo?

Quy commented on plugins/system/debug/DataCollector/ProfileCollector.php:
```diff
@@ -0,0 +1,312 @@
+	 * @param   \Closure     $closure    A closure.
+	 * @param   string|null  $collector  A collector.
+	 *
+	 * @since __DEPLOY_VERSION__
```
align second column

Quy commented on plugins/system/debug/DataCollector/ProfileCollector.php:
```diff
@@ -0,0 +1,312 @@
+	/**
+	 * Returns an array of all measures
+	 *
+	 * @since __DEPLOY_VERSION__
```
align second column. Do the same below.

elkuku commented:
Hello and sorry for the late reply. I had a little accident on my motor bike, broke my finger and so I'm typing a little slow - please be patient :wink:

Let me respond @roland-d :
1. This is not implemented yet (code exists only local). The idea is to display all the request informations with all the bells and whistles that we have on a dedicated page.
1. What Michael said..
1. Yeah because I'm just dumping the array here - Will be pretified in 1.
1. See 1.
1. Where's 5.? :tongue:
1. See 1.
1. I was looking into making those clickable, but I *think* we should leave this also for 1. - please :wink:
1. That looks really bad you're right. On the other hand: If you look closely where the link points to you'll not want click on it...
1. AHA excellent question `*BigGrin*` - There is a "hidden feature" called deprecation logging, I believe most devs are not aware of this...
Please find the plugin setting and.. See my screenshot ;)
![bildschirmfoto_2018-07-31_11-31-28](https://user-images.githubusercontent.com/33978/43473851-c9a9da5e-94b6-11e8-8c45-e4bef924a2e6.png)

1. I'll do my best - Thanks!

On the other hand I removed all the vendor code, so code reviews may be possible now..

I also modified the `build.js` file to copy the assets from the PHP package. This would obviously mean that you have to run `composer install` **before** `npm install` but I guess, since the projects "main" language is PHP, that should be the case anyway...

Suggestions?

elkuku committed:
Fix file path not readable

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

# Conflicts:
#	build/build-modules-js/update.js

elkuku committed:
I like to move it, move it :man_dancing:

elkuku commented:
Travis is green but the words `Segmentation fault` don't have a quite positive sound...

elkuku committed:
Add a @todo

dgrammatiko commented on build/build-modules-js/update.js:
```diff
@@ -16,6 +16,9 @@ const cleanVendors = () => {
   fsExtra.copySync(Path.join(rootPath, 'build/media/vendor/jquery-ui'), Path.join(rootPath, 'media/vendor/jquery-ui'));
 
+  // And here some assets from a PHP package
+  fsExtra.copySync(Path.join(rootPath, 'libraries/vendor/maximebf/debugbar/src/DebugBar/Resources'), Path.join(rootPath, 'media/vendor/debugbar'));
```
We might need to use one of the other methods in this file. The reason is that we want to keep the convention `vendor/name/js`. Mind the last `js` folder. This enables joomla's HTMLHelper::script() to override the scripts if one is found in the relevant template folder (eg `templates/some-template/js/vendor/name/some-script.js`).

Just add a @todo above the line and I will do the rest once this is merged!

elkuku replied:
If it's just about the `/js` folder I guess I could just add that to the `dest` path of the copy command and adjust my path?
I am more concerned about the fact that somebody could run `npm i` **before** running `composer install` so the copy command here would obviously fail...
It just feels that stuff is done in the `wrong_place`&trade;  

In the mean time I have added a `@todo` and feel good having that working (again)

Thanks for your help and keep up the good work :man_dancing:

elkuku committed:
Add a debug view to com_content to display extended information
Oooooo

elkuku committed:
Fix a doc block

elkuku committed:
@todo Move it the 'right way' (tm) :man_dancing:

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

Hackwar commented:
Going to test this now, but I first glanced over the code of the PR and I was wondering why there is a debug model in com_content?

Hackwar commented:
After merging this branch into my smart search changes: I really like this! The display of session data is great and the whole display of the debug data is :+1: Can I somehow help you guys or do you have all the code already locally and just need to clean it up?

elkuku committed:
Remove proposal

elkuku committed:
Use full paths

elkuku committed:
Submit more data

elkuku committed:
Cleanup debug.php
Alot...

elkuku committed:
Fix CS

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

elkuku committed:
Call $db->disconnect() in the "right" place

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

# Conflicts:
#	composer.lock

elkuku committed:
Update composer.lock file

elkuku committed:
Fix CS

elkuku commented:
Hi @Hackwar alles klar? :wink: first of all, thanks for your kind words. Very much appreciated.
Now - the debug model you discovered in `com_content` is part of a proposal that I currently can not find an appropriate place for, so I hacked it in `com_content`. I feel that introducing a new component like `com_debug` or something would be something that need more "fighting with the gate keepers".... So here is the deal:

1. The current branch contains a cleanup of the debug plugin (That is: all HTML code removed, all data processing is done by proper classes, rendering on the current page is done in JavaScript)
1. The removed code has been placed in HTML related classes and views in `com_content` in a separate branch here: https://github.com/elkuku/joomla-cms/tree/com_debug as a proposal and for testing.
Functionality should be the same as in the "old" debug console, or even better (@roland-d Language errors are now displayed one link per error and with the correct line number :wink: )

Happy testing

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

Quy commented on plugins/system/debug/debug.xml:
```diff
@@ -217,6 +168,55 @@
 				label="PLG_DEBUG_LOGGING_FIELDSET_LABEL"
 				>
+				<field
+					name="logs"
```
Per [Joomla Coding Standards](https://developer.joomla.org/coding-standards/xml.html):

> The four attributes `name`, `type`, `label` and `description` should be written in this order and at the top of the element definition.

elkuku replied:
Not sure what you mean here either :disappointed: Most of the "changes" only have been moved from other places without changes.
Please clarify :wink:

Quy replied:
Move the `class` attribute after the `label` attribute. Repeat below.

elkuku replied:
Please have a look at https://github.com/joomla/joomla-cms/blob/4.0-dev/plugins/system/debug/debug.xml
It's just like this...
But I'll change it - no prob..

elkuku replied:
Done.

Quy commented on plugins/system/debug/AbstractDataCollector.php:
```diff
@@ -0,0 +1,79 @@
+abstract class AbstractDataCollector extends DataCollector implements Renderable
+{
+	/**
+	 * @var Registry
```
Please follow format standards for Class Property DocBlocks: https://developer.joomla.org/coding-standards/docblocks.html

elkuku replied:
I went over all the doc blocks again - could you recheck please - Thanks!

elkuku committed:
Review doc blocks - Thanks @Quy

elkuku commented:
System test failures don't seem related to the code changed here...

elkuku committed:
Re-order XML tags - Thanks @Quy

elkuku committed:
Collect only relevant information on untranslated language strings

elkuku commented:
@roland-d there is a PR to fix the untranslated strings display. Maybe you have time to test #21457?

roland-d commented:
@elkuku Do we need the composer.lock in this commit? Because of this file `git apply` won't work.

elkuku commented:
> Do we need the composer.lock in this commit

You mean here in this PR? I'd say yes, otherwise `composer install` won't work as expected?

roland-d commented:
Ok, I will test it in another way.

elkuku committed:
Use the SQLQueriesWidget to display database queries

elkuku committed:
Make loaded language files clickable

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

roland-d commented:
@elkuku Thank you for the update. The clickable languages is not working for me. Here is what I see:
<img width="289" alt="image" src="https://user-images.githubusercontent.com/359377/43989041-d48355ba-9d42-11e8-970d-341841bf2e69.png">

<img width="288" alt="image" src="https://user-images.githubusercontent.com/359377/43989036-c7f1a1a8-9d42-11e8-94e6-02633646184c.png">

Seems like a parsing error or something?

As for the DB queries, looking nice especially that you can click the query to get it copied. I just wonder if we can be notified without a JS alert. 

Another question is, can we filter so we can see which duplicate queries exist?

elkuku commented:
> The clickable languages is not working for me

This happens if an "old" version if the js file is in use. Please make sure to run `npm i` and clear all your browser caches :wink: 

> looking nice especially that you can click the query to get it copied. I just wonder if we can be notified without a JS alert.

Unfortunately we can not modify (or even remove) this functionality without "forking" the source code - which I am very attempted to, just to implement the stack trace display...

> can we filter so we can see which duplicate queries exist?

Same as above - you would have to group them first, display only one of the dupes etc.. Maybe you could add something to https://github.com/maximebf/php-debugbar/issues/382

elkuku committed:
Namespace debug monitor

elkuku committed:
Set a dummy monitor for extended db debugging

roland-d commented:
@elkuku I cleaned things up and it is looking much better now. 

> Unfortunately we can not modify (or even remove) this functionality without "forking" the source code - which I am very attempted to, just to implement the stack trace display...

I was playing around with the Nette framework and they have Tracy the debug bar. https://github.com/nette/tracy for more info. This is also very neat.

> Maybe you could add something to maximebf/php-debugbar#382

I will have a look at it.

The following issue I still observe
<img width="593" alt="image" src="https://user-images.githubusercontent.com/359377/44034279-ff0d59f4-9f0c-11e8-9d32-2f185e6ef941.png">

That should be fixed by the other PR right?

Where in Joomla can you get those Deprecated notices to show up?

elkuku committed:
Add a "fork" of the sqlqueries widget

elkuku committed:
Modify the sqlqueries widget

elkuku committed:
Add a method to format caller information

elkuku committed:
Add stack trace

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

elkuku commented:
Another weekend, another round of features :wink: 
First of all, I "forked" the sqlqueries widget to our code base to perform some hacking.
Currently added fatures:
1. A (clickable) call stack with indication of the "caller" (somewhat experimental)
![joomla 4 debug bar -call-stack](https://user-images.githubusercontent.com/33978/44042430-dc38f482-9ee5-11e8-91ec-9bd2667c322a.png)
1. Duplicated queries can be filtered
![menus items main menu joomla 4 debug bar administration](https://user-images.githubusercontent.com/33978/44042438-dfd1afd0-9ee5-11e8-9770-2dcbccc82443.png)
**Note** the filter buttons are at the lower right corner :wink: 

**NOTE for Re-testers** please execute `npm run update` or `npm i`

@roland-d 

> I was playing around with the Nette framework and they have Tracy the debug bar.

I had a look at tracy before starting here but decided to use phpdebugbar for architectural reasons :wink: 

> The following issue I still observe

Yes the correct functionality depends on the other PR being merged - thanks for testing btw :wink: 

> Where in Joomla can you get those Deprecated notices to show up?

They are all over the place, especially the "core" deprecation, that's why I decided to "filter them... Do you mean you don't get any of them??

elkuku committed:
Fix dupe filter labels

elkuku committed:
Move custom widgets to /media

roland-d commented:
@elkuku 
> Note the filter buttons are at the lower right corner
Awesome. I see for example *Duplicate 1*, let's say I have 30 duplicate queries. Do I get 30 buttons in that case? (assuming these are 30 unique duplicate queries).

> They are all over the place, especially the "core" deprecation, that's why I decided to "filter them... Do you mean you don't get any of them??

Correct I don't get any of them.

<img width="670" alt="image" src="https://user-images.githubusercontent.com/359377/44049346-1e11dde4-9f34-11e8-9e52-f6c6f8af16dd.png">
Can we make those links clickable too?

elkuku commented:
> Correct I don't get any of them.

Do you at least see the "Deprecated" Tab in the debug bar?
Are you sure you enabled `Log Deprecated API` in the plugin parameters as in my screenshot in https://github.com/joomla/joomla-cms/pull/20380#issuecomment-409288517 ?

> Do I get 30 buttons in that case? (assuming these are 30 unique duplicate queries).

How can a duplicated query be unique? :wink: 
The duplicated queries are "grouped" so you will get a button for every "group"
E.g. for the screenshot you posted earlier https://github.com/joomla/joomla-cms/pull/20380#issuecomment-408345262 at point 6. you would get three buttons.
BTW could you do a test with this example? That'd be cool :wink: 

> Can we make those links clickable too?

For now I guess that you should open the call stack and make sure that the "caller" is correct, not sure that this works 100% reliable.. - but nice idea, I'll have a look at this.

roland-d commented:
@elkuku 
> Are you sure you enabled Log Deprecated API in the plugin parameters as in my screenshot in #20380 (comment) ?

It was not enabled. I did look for it before but couldn't find it. The tabs are hard to see in J4. Now I see them all.

> Log Executed Queries
This option was not set but I can see the queries. So I guess that was confusing as well.

> BTW could you do a test with this example?
I will test this and update with the results

roland-d commented:
I set up 3 duplicate queries and here are the results of showing all duplicates

<img width="1435" alt="image" src="https://user-images.githubusercontent.com/359377/44052685-8db41f28-9f3d-11e8-8cdb-2e40c86f88c7.png">

Duplicates 1 turned off

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/359377/44052715-a866bbd2-9f3d-11e8-8545-e884b1bc62e0.png">

Duplicates 2 turned off

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/359377/44052743-b9ee7d68-9f3d-11e8-8d51-4eab05c22bb5.png">

The current buttons feel a bit awkward as in I need to click so much. Something like a toggle all on/off would be nice. I could just toggle all off and then the ones I would like to see. What do you think of that?

elkuku committed:
Make the caller clickable

elkuku committed:
Make the blue bluer

elkuku commented:
> The tabs are hard to see in J4

Fully agree :+1: 

> This option was not set but I can see the queries. So I guess that was confusing as well.

This one actually logs the queries to a file (I believe this has been disabled so the option is useless anyway - will check)

> The current buttons feel a bit awkward as in I need to click so much

Yes, well, I believe if you have that much duplicated queries you should better check your code :yum: 
Actually this is the standard behaviour which is also used to filter log messages.

Meanwhile I made the caller text clickable. Can you confirm those are correct? According to your screenshots you are performing db queries in a template file :wink:

roland-d commented:
@elkuku I tested the caller text clickable for queries and they work but there is a weird side-effect. Every time I click on the link it opens in PhpStorm as expected but it also opens a blank tab. This only happens with the links in the queries.

> Actually this is the standard behaviour which is also used to filter log messages.

So we can't have a button to toggle them all on/off?

elkuku committed:
Remove target attribute from caller link

elkuku commented:
> it also opens a blank tab

I *think* this should be fixed now - please test.
May I ask how you have set up those clickable links? What's your OS and browser?
I have to confess that this never worked reliable for me, e.g. currently I am on a linux box using chrome or firefox and the most "easy to setup" option was to define the xdebug link as `http://localhost:63342/api/file/%f:%l` which works since phpstorm listens on this port while it runs and you can do a lot of nifty things but - I guess the browser will always try to open a new page/tab :(

> So we can't have a button to toggle them all on/off?

I didn't say we can't
It's "our" code so the only thing we need to implement a feature is to find somebody with some time to do it

BTW: You complained about the annoying alert message that appears when copying a query... 
The code is now under "our control" can be found here https://github.com/joomla/joomla-cms/pull/20380/files#diff-704b78076b8b57b6b9b63fa76f3fd570R44 open for modification :wink:

elkuku committed:
Change an icon

elkuku committed:
aDD SOME j! info

elkuku committed:
Fix a parse error

elkuku committed:
Fix CS

roland-d commented:
@elkuku 
> It's "our" code so the only thing we need to implement a feature is to find somebody with some time to do it

Good to know, I will check when I have some time, the same goes for the alert :)

<img width="1102" alt="image" src="https://user-images.githubusercontent.com/359377/44250242-e1125500-a1f3-11e8-8755-b9c22bb0a915.png">

We have another caller here, can you make that clickable as well? The same goes for the Deprecated tab.

<img width="983" alt="image" src="https://user-images.githubusercontent.com/359377/44250290-0901b880-a1f4-11e8-9a8f-4a5ae55b3098.png">

Is it possible to get a caller here?

An inconsistency I noticed is that I have no entries in the Log and Untranslated tab and no entries in the Errors tab. The log tab is completely empty but the Errors tab says 
<img width="314" alt="image" src="https://user-images.githubusercontent.com/359377/44250340-38b0c080-a1f4-11e8-9ab7-ce2c3398fa80.png">

I think they should just be blank as the Log and Untranslated tab is if there is no issue.

<img width="260" alt="image" src="https://user-images.githubusercontent.com/359377/44250479-c7bdd880-a1f4-11e8-979b-f8f9d2d859a6.png">

The Info link takes me to `http://joomlacms.test/index.php?option=com_content&view=debug&id=X862e7a4b7e8180b879610361840e5995` I am not sure what to expect from this but now it gives me a 404.

elkuku committed:
Make query traces and request data optional

elkuku committed:
Display a blank tab if there are no language errors

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

# Conflicts:
#	composer.lock

elkuku committed:
Remove unused import

elkuku committed:
Update composer.lock file

elkuku commented:
> We have another caller here, can you make that clickable as well? The same goes for the Deprecated tab.

I am going to look into this but it will require a fair amount of work so I'm not sure if we should do it after (and only if) this gets merged...

> is it possible to get a caller here?

No - Those are just "messages" with everything that is deprecated.

> but the Errors tab says

Has been fixed - thanks!

> The Info link takes me to

This is supposed to lead you to a page with extended information... The code is actually here: https://github.com/elkuku/joomla-cms/tree/com_debug

BTW I opened a PR regarding the language error display - If you have time to test #21840 that'd be cool.

Thanks for all your testing :wink:

roland-d commented:
@elkuku 
![image](https://user-images.githubusercontent.com/359377/44615526-41307900-a83c-11e8-98d5-3b0f997ece17.png)

I noticed now that these lines are ended with an ellipses but there is more than enough space on the line. Can we have the text just run to the end of the line?

![image](https://user-images.githubusercontent.com/359377/44615547-cb78dd00-a83c-11e8-8f3d-97742eae7980.png)

Can we have a counter behind the number of loaded language files?

> This is supposed to lead you to a page with extended information... The code is actually here: 

Let me know when it is ready to test.

> Thanks for all your testing 😉

Thank you for the coding. I would love to see this go into 4.0. That will be a great help.

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

# Conflicts:
#	plugins/system/debug/debug.php

elkuku committed:
Add a badge for loaded language files

elkuku committed:
Make language errors clickable

elkuku commented:
> I noticed now that these lines are ended with an ellipses but there is more than enough space on the line. Can we have the text just run to the end of the line?

Your output will be different on a smaller screen - I'd leave this for now



> Can we have a counter behind the number of loaded language files?

Done



> Let me know when it is ready to test.

I think when this got merged I'll get into bringing this to a "testable" state...

elkuku committed:
Remove commented code

roland-d commented:
@elkuku I had another look, this looks good to me. For me this is good enough to merge, we just need to remove the *info* link because that doesn't do anything. That can be added later once the component is ready.

brianteeman commented on plugins/system/debug/debug.xml:
```diff
@@ -20,6 +20,18 @@
+					description="PLG_DEBUG_FIELD_REFRESH_ASSETS_DESC"
+					class="switcher"
+					default="1"
+				>
```
please indent one more tab

brianteeman commented on plugins/system/debug/debug.xml:
```diff
@@ -29,119 +41,82 @@
-					label="PLG_DEBUG_FIELD_PROFILING_LABEL"
 					default="1"
-					>
+				>
```
please revert this change it was correct before

elkuku replied:
Done - i guess...

elkuku committed:
Format XML file

elkuku committed:
Disable info link

elkuku commented:
> we just need to remove the info link because that doesn't do anything.

Right - has been removed.

franz-wohlkoenig commented:
> when this got merged

should Status set be on "RTC"?

laoneo commented:
This should get another round of tests, before setting to RTC.

franz-wohlkoenig commented:
Thanks for Info @laoneo that this Pull Request needs also 2 successfully Tests.

brianteeman commented:
Thank you for this @elkuku it looks really good and I am sure it will become a great addition. 

Would it be possible to 
- [ ] change the php logo with administrator/templates/atum/images/logo-blue.svg
- [ ] add mysql version to the J! Info section
- [ ] have some way of clearly identifying the language related parts - currently its not obvious until you click on them
![image](https://user-images.githubusercontent.com/1296369/44709229-024f2d00-aaa1-11e8-9efd-9f45ddd739c8.png)
- [ ] be consistent in the fonts and styling used between tabs. 
![image](https://user-images.githubusercontent.com/1296369/44709425-80abcf00-aaa1-11e8-91fd-469437a6df20.png)
![image](https://user-images.githubusercontent.com/1296369/44709440-8e615480-aaa1-11e8-8981-dc91a0746c61.png)

brianteeman commented on build/media/plg_system_debug/widgets/info/widget.js:
```diff
@@ -0,0 +1,56 @@
+                tr = $('<tr />')
+                    .append($('<td />').text('Template'))
+                    .append($('<td />').text(data.template.template));
+                    // .append($('<td />').html('<ul><li>huhu</li><li>haha</li></ul>'));
```
Looks like this is some testing code that should be removed

elkuku replied:
Oops..
I've left however some commented code with a `@todo`. I hope this is acceptable.

elkuku committed:
Remove obsolete CSS

elkuku committed:
Change debugbar logo

elkuku committed:
Add database information

elkuku committed:
Remove obsolete code

elkuku committed:
Fix CS

brianteeman commented:
Thanks for making those changes. I saw the @todo and thats fine for me - don't know about others

I was just testing a duplicate query issue and it looked like this debugger was giving a different number of queries on the control panel page than the current debugger???

elkuku commented:
> change the php logo with administrator/templates/atum/images/logo-blue.svg

So today I learned how to embed an svg in a css file. It wasn't just copy&paste but involved some hacking (e.g. hex color codes are not supported...) but in the end there is a Joomla! logo
![bildschirmfoto_2018-08-28_12-48-23](https://user-images.githubusercontent.com/33978/44741527-f70cf900-aac2-11e8-9c53-b1a686fe76e7.png)

> add mysql version to the J! Info section

Done, see screenshot above. 
I am using the `SysInfoModel` from `com_admin` here (why is this class still not autoloadable with a dozen autoloaders in place?? :cry: ) so we have pretty much every information ready for display as in `backend => sysinfo`

> have some way of clearly identifying the language related parts - currently its not obvious until you click on them

I looked into adding a tooltip at least but this doesn't seem to be supported. The other option would be to rename the tab title to make it clearer, but I feel that they might become pretty "long". Maybe we should use abbreviations like `Lang Loaded` and `Lang Errors`? - WDYT

> be consistent in the fonts and styling used between tabs.

The `Untranslated Language Strings` (Should we really put this string in the title?...) tab will receive its "own tab" (and corresponding styling) after #21457  :wink:

elkuku commented:
> I was just testing a duplicate query issue and it looked like this debugger was giving a different number of queries on the control panel page than the current debugger???

That shouldn't be possible since it is using the exact same data - thinking - do you mean the number of duplicates or the total number? both should be the same... `need more info` :wink:

brianteeman commented:
I will send more info tomorrow. Not at my pc right now. For the SVG you can add colour to it with css

brianteeman commented:
Clean install of J4 with multilingual sample data and testing administrator/index.php
With debug on it produced
```
28 Queries Logged 37.78 ms

4 duplicate found!
2 duplicates: #17  #20
2 duplicates: #40  #43
```
With this new debug on it produced

`20 statements were executed, 2 of which were duplicates, 18 unique`

elkuku commented:
Umm the only thing I can think of is that there are additional changes in
the 4.0 branch which are producing different queries.
I'll update the branch and do some testing later.

Brian Teeman <notifications@github.com> schrieb am Mi., 29. Aug. 2018,
06:30:

> Clean install of J4 with multilingual sample data and testing
> administrator/index.php
> With debug on it produced
>
> 28 Queries Logged 37.78 ms
>
> 4 duplicate found!
> 2 duplicates: #17  #20
> 2 duplicates: #40  #43
>
> With this new debug on it produced
>
> 20 statements were executed, 2 of which were duplicates, 18 unique
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/20380#issuecomment-416919917>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACEukuAjMd7X9AGqjMYZoa3xo1GSzT7ks5uVntagaJpZM4T8kvb>
> .
>

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

elkuku commented:
Sorry yesterday was to short... I would be in favor not only to abolish this crazy DST thing but also to add some additional hours to every day so we can be more productive...

That said, here are my findings:
I checked out the latest from 4.0, installed the german language pack, installed ML sample data, sailed to admin/index.php and got - nothing :disappointed: no duplicated queries...

Then I switched to the debug-bar branch and - got some dupes:
![bildschirmfoto_2018-08-31_13-23-02](https://user-images.githubusercontent.com/33978/44929868-15bcfb00-ad22-11e8-8af7-ca608a0cb763.png)
Then I merged in the 4.0 branch and - no more dupes...

So I guess I was right with my previous assumption. IIRC there was a PR about duplicated queries merged recently - could you test once more please?

Hackwar commented:
I tested this and this looks very good. I can't speak on the missing queries part, but what I'm missing is the used memory of the different parts in the Profile tab. Can you have a look there?

I also have my issues with the design. I guess it is mostly due to being accustomed to the current style of the debug bar. Right now the texts to me appear pretty much like a wall. Especially the queries are not easy to read for me. Maybe we can color the icons/buttons/action areas of each query in the Joomla colors? And have a stronger seperator between the queries?

infograf768 commented:
Class 'DebugBar\DebugBar' not found

elkuku commented:
Please run composer install and npm install

infograf768 <notifications@github.com> schrieb am Fr., 31. Aug. 2018, 23:53:

> Class 'DebugBar\DebugBar' not found
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/20380#issuecomment-417833031>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACEupIi3dtnAx78Czf24UBi9_mZ2npYks5uWhLNgaJpZM4T8kvb>
> .
>

infograf768 commented:
Yeah, I was in the wrong repository.
Had an error though. Had to type `y`

```
Updating joomla/test-system dev-4.0-dev (472c69a => 0548840):     The package has modified files:
    M src/drone-run.sh
    Discard changes [y,n,v,d,s,?]? y
```

infograf768 commented:
Needs some kind work. 😃 
UI is very confusing. Things are stuck to one another and hardly readable (font size, line height). 


Here color instead of Loaded/Not Loaded (not good for colour blind people). Maybe using both would be better.

<img width="810" alt="screen shot 2018-09-01 at 07 22 00" src="https://user-images.githubusercontent.com/869724/44942690-16a06c00-adb8-11e8-9b01-f5150c2c9c2a.png">

Here constants are not displayed fully.

<img width="717" alt="screen shot 2018-09-01 at 07 22 14" src="https://user-images.githubusercontent.com/869724/44942692-1ef8a700-adb8-11e8-814f-79ec1089b9fc.png">

NOTE: just found out a change in Loaded/Not Loaded ini files. Must have been for some times as it is the same in 3.8.x
When a language file does not exist in a language, it is rightfully tagged as Not Loaded, but if I remember well, some time ago, it would show nearby the en-GB equivalent file as loaded instead.

elkuku committed:
Add a visual indicator for loaded language files - thanks @infograf768

elkuku committed:
Display accurate location and a stack trace for untranslated language strings

elkuku committed:
Add more return types

elkuku committed:
Add more return types

elkuku committed:
Fix a path

elkuku committed:
Fix a notice

elkuku committed:
Make the separator line blue and add some padding to query list

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

elkuku commented:
@Hackwar 
> what I'm missing is the used memory of the different parts in the Profile tab. Can you have a look there?

I think we can at least display the amount (using numbers) would this be helpful?

> the queries are not easy to read for me

I've changed the color of the separator line to a beautiful blue and added some padding - what do you think/see?

![bildschirmfoto_2018-09-02_12-34-16](https://user-images.githubusercontent.com/33978/44958911-9dcf0c00-aeac-11e8-9dd9-312d897784cd.png)

> Maybe we can color the icons/buttons/action areas of each query in the Joomla colors?

That sounds like a nice idea but I'm afraid that if I perform this task it will look horrible - trust me :tongue: 

@infograf768 

> Here color instead of Loaded/Not Loaded (not good for colour blind people). Maybe using both would be better. 

I've added a `+` or a `-` sign accordingly - WDYT?

> Alsdo no need imho to repeat the name of the file with these 2 columns.

The first column is supposed to be the "extension" that is trying to load the language file(s) - Note the first line `joomla` is loading the `xx-XX.ini` file :wink: I've also added a header to make this clearer. I believe this is nice for "grouping" the files.

> if I remember well, some time ago, it would show nearby the en-GB equivalent file as loaded instead.

Actually I noticed the same thing. As you can see in my screenshot this is different in admin and site...

![bildschirmfoto_2018-09-02_11-08-43](https://user-images.githubusercontent.com/33978/44958113-fac4c500-aea0-11e8-870b-46f0d0c2812e.png)

> Here constants are not displayed fully.

So instead of waiting for #21457 being merged, I changed (J)Language here in this PR and also added the full stack trace.

![bildschirmfoto_2018-09-02_11-09-44](https://user-images.githubusercontent.com/33978/44958114-fb5d5b80-aea0-11e8-8992-ef1644259c0c.png)

laoneo commented:
I had a look on your pr and some parts are missing according to your description. For example the duplicates query or the call stack of a query.

elkuku commented:
> ome parts are missing according to your description. For example the duplicates query or the call stack of a query.

I made the query traces optional to avoid performance problems on "large" sites. so please make sure that you enabled them in plugin options.

![plugins system debug j4 debugbar administration](https://user-images.githubusercontent.com/33978/44994295-9b7bb900-af63-11e8-88f3-8c9534184cc3.png)

Note: The "Query Types" option is currently not functional

As for the duplicated queries, that should work as announced - what output do you get / expect?

elkuku committed:
Add memory marks on profiler tab = thanks @Hackwar

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

elkuku committed:
Remove unused parameter

elkuku committed:
Make quwey profiles and explains optional
Remove unused query log option

elkuku committed:
Update language file

elkuku committed:
Add support for query explains

elkuku commented:
Just a small update for possible testers... Support for query explains has been added.
![bildschirmfoto_2018-09-04_11-39-58](https://user-images.githubusercontent.com/33978/45045025-722c5d00-b037-11e8-9f98-2d914a30f273.png)

elkuku committed:
Merge remote-tracking branch 'origin/4.0-dev' into debug-bar-2

elkuku committed:
Modify CSS

laoneo commented:
Can we get one more test and then I would suggest we merge it and do finalize the other issues in their own pr's.

roland-d commented:
@elkuku Great, great work. Well done.
@laoneo Here is my second test again. The things is awesome works great.

I have some ideas for enhancements but let's do that in a new issues.

laoneo commented:
Thank you very much for this nice change!

roland-d commented:
Very happy to see this go in. +1 for progress.

elkuku commented:
:tada:
